README.md
# Brainard


ChangeLog.md
# Changelog for Brainard

## Unreleased changes


test/Brainard/ModelSpec.hs
{-# LANGUAGE OverloadedStrings #-}

module Brainard.ModelSpec where

import Brainard.Model
import Common
import Conduit (ResourceT)
import Control.Monad.IO.Unlift (MonadUnliftIO)
import Data.Time.Calendar.OrdinalDate (fromOrdinalDate)
import Database.Esqueleto.Experimental
import qualified Database.Persist.Sqlite as S
import Test.Hspec


sqlite :: ReaderT SqlBackend (NoLoggingT (ResourceT IO)) a -> IO a
sqlite action = S.runSqlite ":memory:" $ do
  runMigrationSilent migrateAll
  action

time :: Integer -> UTCTime
time n = UTCTime (fromOrdinalDate 2022 1) (secondsToDiffTime n)

spec :: SpecWith ()
spec = do
  describe "Rememberance Crud" $ do
    it "Can Be Added" $
      sqlite $ do
        postRememberance "A" (Just $ time 0)
        rs' <- getRememberances
        let rs = map entityVal rs'
        liftIO $ rs `shouldBe` [Rememberance "A" (time 0) 0 0]
    it "Orders by Date" $ do
      sqlite $ do
        postRememberance "A" (Just $ time 0)
        postRememberance "B" (Just $ time 1)
        rs' <- getRememberances 
        let rs = map (rememberanceBody . entityVal) rs'
        liftIO $ rs `shouldBe` ["B", "A"]
    it "Can Upvote" $ do
      sqlite $ do
        postRememberance "A" (Just $ time 0)
        rs' <- getRememberances 
        let k = entityKey . head $ rs'
        upvote k
        rs' <- getRememberances
        let upvotes = rememberanceUpvotes . entityVal . head $ rs'
        liftIO $ upvotes `shouldBe` 1
    it "Can Downvote" $ do
      sqlite $ do
        postRememberance "A" (Just $ time 0)
        rs' <- getRememberances 
        let k = entityKey . head $ rs'
        downvote k
        rs' <- getRememberances
        let downvotes = rememberanceDownvotes . entityVal . head $ rs'
        liftIO $ downvotes `shouldBe` 1
  describe "Rememberance Scores" $ do
    it "Upvotes Increase Score" $ do
      sqlite $ do
        postRememberance "A" (Just $ time 0)
        postRememberance "B" (Just $ time 0)
        rs' <- getRememberances
        let [a, b] = map entityKey rs'
        upvote a
        rs' <- getRememberancesByScorePage 2 0 (Just $ time 1)
        let favoredKey = entityKey . head $ rs'
        liftIO $ a `shouldBe` favoredKey
    it "Downvote Decreases Score" $ do
      sqlite $ do
        postRememberance "A" (Just $ time 0)
        postRememberance "B" (Just $ time 0)
        rs' <- getRememberances
        let [a, b] = map entityKey rs'
        downvote b
        rs' <- getRememberancesByScorePage 2 0 (Just $ time 1)
        let favoredKey = entityKey . head $ rs'
        liftIO $ a `shouldBe` favoredKey
    it "Age Decreases Score" $ do
      sqlite $ do
        postRememberance "A" (Just $ time 100)
        postRememberance "B" (Just $ time 0)
        rs' <- getRememberances
        let [a, b] = map entityKey rs'
        downvote b
        rs' <- getRememberancesByScorePage 2 0 (Just $ time 100)
        let favoredKey = entityKey . head $ rs'
        liftIO $ a `shouldBe` favoredKey

ts/src/main.ts
interface KVRememberance {
  key: number,
  val: Rememberance
}

interface Rememberance {
body: string,
upvotes: number,
downvotes: number
}

function getRememberances (limit : number, offset : number, 
  onSuccess : (a?: KVRememberance[]) => void , onError : (e: any) => void) : void {
  var xhr = new XMLHttpRequest();
  xhr.open('GET', '/rememberances' 
    + '?limit=' + encodeURIComponent(Number(limit)) 
    + '&offset=' + encodeURIComponent(Number(offset)), true);
  xhr.setRequestHeader('Accept', 'application/json');
  xhr.onreadystatechange = function () {
      var res = null;
      if (xhr.readyState === 4) {
          if (xhr.status === 204 || xhr.status === 205) {
              onSuccess();
          }
          else if (xhr.status >= 200 && xhr.status < 300) {
              try {
                  res = JSON.parse(xhr.responseText);
              }
              catch (e) {
                  onError(e);
              }
              if (res)
                  onSuccess(res);
          }
          else {
              try {
                  res = JSON.parse(xhr.responseText);
              }
              catch (e) {
                  onError(e);
              }
              if (res)
                  onError(res);
          }
      }
};
xhr.send(null);
};
function postRememberances (body: string, 
    onSuccess : (a?: number) => void, onError : (e: any) => void) : void {
  var xhr = new XMLHttpRequest();
  xhr.open('POST', '/rememberances', true);
  xhr.setRequestHeader('Accept', 'application/json');
  xhr.setRequestHeader('Content-Type', 'application/json');
  xhr.onreadystatechange = function () {
      var res = null;
      if (xhr.readyState === 4) {
          if (xhr.status === 204 || xhr.status === 205) {
              onSuccess();
          }
          else if (xhr.status >= 200 && xhr.status < 300) {
              try {
                  res = JSON.parse(xhr.responseText);
              }
              catch (e) {
                  onError(e);
              }
              if (res)
                  onSuccess(res);
          }
          else {
              try {
                  res = JSON.parse(xhr.responseText);
              }
              catch (e) {
                  onError(e);
              }
              if (res)
                  onError(res);
          }
      }
};
xhr.send(JSON.stringify(body));
};
function postRememberancesByIdUpvote (id: number, 
  onSuccess : (a?: any) => void, onError : (e: any) => void) : void {
var xhr = new XMLHttpRequest();
xhr.open('POST', '/rememberances/' + encodeURIComponent(id.toString()) + '/upvote', true);
xhr.setRequestHeader('Accept', 'application/json');
xhr.onreadystatechange = function () {
    var res = null;
    if (xhr.readyState === 4) {
        if (xhr.status === 204 || xhr.status === 205) {
            onSuccess();
        }
        else if (xhr.status >= 200 && xhr.status < 300) {
            try {
                res = JSON.parse(xhr.responseText);
            }
            catch (e) {
                onError(e);
            }
            if (res)
                onSuccess(res);
        }
        else {
            try {
                res = JSON.parse(xhr.responseText);
            }
            catch (e) {
                onError(e);
            }
            if (res)
                onError(res);
        }
    }
};
xhr.send(null);
};
function postRememberancesByIdDownvote (id : number, 
  onSuccess : (a? : any) => void, onError : (e:any) => void) : void {
  var xhr = new XMLHttpRequest();
  xhr.open('POST', '/rememberances/' + encodeURIComponent(id.toString()) + '/downvote', true);
  xhr.setRequestHeader('Accept', 'application/json');
  xhr.onreadystatechange = function () {
      var res = null;
      if (xhr.readyState === 4) {
          if (xhr.status === 204 || xhr.status === 205) {
              onSuccess();
          }
          else if (xhr.status >= 200 && xhr.status < 300) {
              try {
                  res = JSON.parse(xhr.responseText);
              }
              catch (e) {
                  onError(e);
              }
              if (res)
                  onSuccess(res);
          }
          else {
              try {
                  res = JSON.parse(xhr.responseText);
              }
              catch (e) {
                  onError(e);
              }
              if (res)
                  onError(res);
          }
      }
};
xhr.send(null);
};

function listRememberances(limit: number, offset: number, container: HTMLElement): void {
  getRememberances(limit, offset, (rs) => {
    rs.forEach((r) => {
      let n = createRememberanceElement(r);
      container.appendChild(n);
    })
  }, (e) => {
    throw new Error(e);
  });
}

function listRememberancesOnto(containerId: string): void {
  let container = document.getElementById(containerId);
  let limit = parseInt(container.getAttribute("data-limit"));
  let offset = parseInt(container.getAttribute("data-offset"));
  listRememberances(limit, offset, container);
}

function createRememberanceElement(r: KVRememberance): HTMLElement {
  let outside = document.createElement('div') as HTMLElement;
  outside.id = "out" + r.key.toString();
  outside.className = 'rememberanceHolder';
  let n = document.createElement('div') as HTMLElement;
  n.id = "r" + r.key.toString();
  n.className = "rememberance";
  let body = document.createElement('div') as HTMLElement;
  body.textContent = r.val.body;
  body.className = "rText"
  n.appendChild(body);
  let upvote = document.createElement('a') as HTMLAnchorElement;
  upvote.id = "r" + r.key.toString() + "Upvote";
  upvote.className = "upvote";
  upvote.textContent = "+" + r.val.upvotes.toString();
  upvote.title = "Click to Upvote";
  upvote.href = "#";
  upvote.onclick = function (e) {
    postRememberancesByIdUpvote(r.key, () => {
      let upvotes = parseInt(upvote.textContent.slice(1));
      upvote.textContent = "+" + (upvotes + 1).toString();
    }, (e) => {throw Error(e);})
  }
  n.appendChild(upvote);
  let downvote = document.createElement('a') as HTMLAnchorElement;
  downvote.id = "r" + r.key.toString() + "Upvote";
  downvote.className = "downvote";
  downvote.textContent = "-" + r.val.downvotes.toString();
  downvote.title = "Click to Downvote";
  downvote.href = "#";
  downvote.onclick = function (e) {
    postRememberancesByIdDownvote(r.key, () => {
      let downvotes = parseInt(downvote.textContent.slice(1));
      downvote.textContent = "-" + (downvotes + 1).toString();
    }, (e) => {throw Error(e);})
  }
  n.appendChild(downvote);
  outside.append(n);
  return outside;
}


function submitRememberanceForm(textBoxId: string, containerId: string) {
  let textBox = document.getElementById(textBoxId) as HTMLInputElement;
  let body = textBox.value;
  let container = document.getElementById(containerId);
  postRememberances(body, (id) => {
    textBox.value = "I remember ";
    let r : KVRememberance = {
      key: id,
      val: {
        body: body,
        upvotes: 0,
        downvotes: 0
      }
    }
    let n = createRememberanceElement(r);
    container.prepend(n);
  }, (e) => {
    throw new Error(e);
  });

}

src/Brainard/Model.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE EmptyDataDecls #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE RankNTypes #-}
module Brainard.Model where

import Database.Esqueleto.Experimental
import Database.Persist.TH (mkMigrate, mkPersist, persistLowerCase, share, sqlSettings)
import Data.Text (Text)
import Common
import Data.Coerce (coerce)
import GHC.Generics (Generic)
import Data.Data (Typeable)
import Data.Swagger (ToSchema)
import Data.Time.Calendar.OrdinalDate (fromOrdinalDate)


share
  [mkPersist sqlSettings, mkMigrate "migrateAll"]
  [persistLowerCase|
  Rememberance json
    body Text sqltype=varchar(512)
    created UTCTime default=CURRENT_TIMESTAMP
    createdEpochSeconds Int64 default=0
    upvotes Int64 default=0
    downvotes Int64 default=0
    deriving Eq Show Generic Typeable
|]

instance ToSchema Rememberance

type InsertContext m b = (MonadIO m, PersistStoreWrite b, BaseBackend b ~ SqlBackend)
type UpdateContext m b = (MonadIO m, BackendCompatible SqlBackend b, PersistQueryWrite b, PersistUniqueWrite b)

timeOrCurrent :: (MonadIO m) => Maybe UTCTime -> m UTCTime
timeOrCurrent (Just x) = return x
timeOrCurrent Nothing = liftIO getCurrentTime

secondsSinceEpoch :: UTCTime -> Int64
secondsSinceEpoch t =
    let x = t `diffUTCTime` UTCTime (fromOrdinalDate 1970 1) (secondsToDiffTime 0)
    in floor $ nominalDiffTimeToSeconds x

postRememberance :: (InsertContext m b) => Text -> Maybe UTCTime -> ReaderT b m (Key Rememberance)
postRememberance body mTime = do
  t <- timeOrCurrent mTime
  let d = secondsSinceEpoch t
  insert (Rememberance body t d 0 0)

getRememberances :: MonadIO m => SqlReadT m [Entity Rememberance]
getRememberances = select $ do
  r <- from $ table @Rememberance
  orderBy [ desc $ r^.RememberanceCreated ]
  pure r

type Limit = Int64
type Offset = Int64

score :: SqlExpr (Entity Rememberance) -> UTCTime -> SqlExpr (Value Double)
score r now = score
  where
    nowS = secondsSinceEpoch now
    -- Idea from:  https://julesjacobs.com/2015/08/17/bayesian-scoring-of-ratings.html
    upvotes :: SqlExpr (Value Double)
    upvotes = castNum  (r ^.RememberanceUpvotes) +. val 4.0
    downvotes :: SqlExpr (Value Double)
    downvotes = castNum (r ^. RememberanceDownvotes) +. val 10.0

    age = castNum (val nowS) -. castNum (r ^. RememberanceCreatedEpochSeconds)
    ageFactor = val 86400  /. (age +. val 1) 
    score = ageFactor *. (upvotes /. (upvotes +. downvotes))

getRememberancesByScorePage :: MonadIO m => Limit -> Offset -> Maybe UTCTime -> SqlReadT m [Entity Rememberance]
getRememberancesByScorePage limitN offsetN mTime =
  map fst <$> getRememberancesByScorePageWithScore limitN offsetN mTime

getRememberancesByScorePageWithScore :: MonadIO m => Limit -> Offset -> Maybe UTCTime -> SqlReadT m [(Entity Rememberance, Value Double)]
getRememberancesByScorePageWithScore limitN offsetN mTime = do
  n <- timeOrCurrent mTime
  select $ do
    r <- from $ table @Rememberance
    orderBy [ desc $ score r n]
    pure (r, score r n)

upvote :: UpdateContext m b => RememberanceId -> ReaderT b m Bool
upvote rememberanceId = do
  n <- updateCount $ \r ->  do
    set r [RememberanceUpvotes =. (r^.RememberanceUpvotes) +. val 1]
    where_ $ r^.RememberanceId ==. val rememberanceId
  return $ n == 1

downvote :: UpdateContext m b => RememberanceId -> ReaderT b m Bool
downvote rememberanceId = do
  n <- updateCount $ \r -> do
    set r [RememberanceDownvotes =. (r^.RememberanceDownvotes) +. val 1]
    where_ $ r^.RememberanceId ==. val rememberanceId
  return $ n == 1


src/Brainard/Server.hs
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
module Brainard.Server where

import Brainard.Model (Rememberance)
import qualified Brainard.Model as M
import Common
import Control.Monad.IO.Unlift (MonadUnliftIO)
import Data.Text (Text)
import Database.Esqueleto.Experimental (BackendCompatible, Entity (Entity), SqlBackend, SqlPersistT, runSqlConn, toSqlKey, runMigrationSilent, runMigration, fromSqlKey, Value (unValue))
import Servant
import Servant.JS
import Servant.Swagger
import Data.Swagger(Swagger, ToSchema, info, title, version)
import Control.Exception (try, throwIO)
import Database.Persist.Sqlite (withSqliteConn)
import Network.Wai.Handler.Warp (run)
import Data.Aeson (ToJSON, ToJSON2 (liftToJSON2), object, KeyValue ((.=)), defaultOptions, genericToEncoding)
import GHC.Generics (Generic)
import Data.Data (Typeable)
import Data.Aeson.Types (ToJSON2(liftToEncoding2), ToJSON (toEncoding))
import Control.Lens

data DbRecord a b = DbRecord { key :: a, val :: b } deriving (Eq, Ord, Show, Generic, Typeable)

instance (ToJSON a, ToJSON b) => ToJSON (DbRecord a b) where
  toEncoding = genericToEncoding defaultOptions

instance (ToSchema a, ToSchema b) => ToSchema (DbRecord a b)

type RememberancesGet = QueryParam "limit" Int64 :> QueryParam "offset" Int64 :> Get '[JSON] [DbRecord Int64 Rememberance]

type PostRememberance = ReqBody '[JSON] Text :> Post '[JSON] Int64

type UpvoteRememberance = "upvote" :> PostNoContent

type DownvoteRememberance = "downvote" :> PostNoContent

type InteractRememberance = Capture "id" Int64 :> (UpvoteRememberance :<|> DownvoteRememberance)

type MainAPI = "rememberances" :> (RememberancesGet :<|> PostRememberance :<|> InteractRememberance)

type API = MainAPI :<|> ("api" :> Get '[JSON] Swagger) :<|> Raw

type AppContext m = (MonadIO m, MonadReader SqlBackend m)

mainAPI :: Proxy MainAPI
mainAPI = Proxy

api :: Proxy API
api = Proxy

apiJs :: FilePath -> IO ()
apiJs = writeJSForAPI mainAPI vanillaJS

apiSwagger :: (Monad m) => m Swagger
apiSwagger = return $ toSwagger mainAPI
  & info.title .~ "Brainard"
  & info.version .~ "1.0"

type APIContext m = (MonadUnliftIO m, MonadReader SqlBackend m)

apiServer :: APIContext m => ServerT API m
apiServer =
  (getRememberances :<|> postRememberance
    :<|> (\id -> upvoteRememberance id :<|> downvoteRememberance id))
  :<|> apiSwagger
  :<|> serveDirectoryWebApp "static"

getRememberances ::
  (APIContext m) =>
  Maybe M.Limit ->
  Maybe M.Offset ->
  m [DbRecord Int64 Rememberance]
getRememberances limitM offsetM =
  let limit = fromMaybe 100 limitM
      offset = fromMaybe 0 offsetM
   in runDb $ do
     rs <- M.getRememberancesByScorePageWithScore limit offset Nothing
     liftIO $ mapM_ (\(_, v) -> print $ unValue v) rs
     let rs' = map (\(Entity k v, _) -> DbRecord (fromSqlKey k) v) rs
     return rs'

postRememberance :: APIContext m => Text -> m Int64
postRememberance body = runDb $ do
  key <- M.postRememberance body Nothing
  return $ fromSqlKey key

upvoteRememberance :: APIContext m => Int64 -> m NoContent
upvoteRememberance id = runDb $ do
  x <- M.upvote . toSqlKey $ id
  unless x . liftIO $ throwIO err404
  return NoContent

downvoteRememberance :: APIContext m => Int64 -> m NoContent
downvoteRememberance id = runDb $ do
  x <- M.downvote . toSqlKey $ id
  unless x . liftIO $ throwIO err404
  return NoContent

type AppT m a = SqlPersistT m a

type App a = AppT IO a

runApp :: MonadIO m => AppT m a -> SqlBackend -> m a
runApp = runReaderT

convertApp :: App a -> SqlBackend -> Handler a
convertApp a = Handler . ExceptT . try . runApp a

runDb :: APIContext m => SqlPersistT m b -> m b
runDb action = ask>>=runSqlConn action

hoistAppServer :: SqlBackend -> Server API
hoistAppServer b = hoistServer api (`convertApp` b) apiServer

app :: SqlBackend -> Application
app = serve api . hoistAppServer

runAppSqliteConn :: Text -> IO ()
runAppSqliteConn conn = runStderrLoggingT $ withSqliteConn conn (\db -> do
  runReaderT (runMigration M.migrateAll) db
  LoggingT (\l -> run 8081 . app $ db))

static/js/main.js
function getRememberances(limit, offset, onSuccess, onError) {
    var xhr = new XMLHttpRequest();
    xhr.open('GET', '/rememberances'
        + '?limit=' + encodeURIComponent(Number(limit))
        + '&offset=' + encodeURIComponent(Number(offset)), true);
    xhr.setRequestHeader('Accept', 'application/json');
    xhr.onreadystatechange = function () {
        var res = null;
        if (xhr.readyState === 4) {
            if (xhr.status === 204 || xhr.status === 205) {
                onSuccess();
            }
            else if (xhr.status >= 200 && xhr.status < 300) {
                try {
                    res = JSON.parse(xhr.responseText);
                }
                catch (e) {
                    onError(e);
                }
                if (res)
                    onSuccess(res);
            }
            else {
                try {
                    res = JSON.parse(xhr.responseText);
                }
                catch (e) {
                    onError(e);
                }
                if (res)
                    onError(res);
            }
        }
    };
    xhr.send(null);
}
;
function postRememberances(body, onSuccess, onError) {
    var xhr = new XMLHttpRequest();
    xhr.open('POST', '/rememberances', true);
    xhr.setRequestHeader('Accept', 'application/json');
    xhr.setRequestHeader('Content-Type', 'application/json');
    xhr.onreadystatechange = function () {
        var res = null;
        if (xhr.readyState === 4) {
            if (xhr.status === 204 || xhr.status === 205) {
                onSuccess();
            }
            else if (xhr.status >= 200 && xhr.status < 300) {
                try {
                    res = JSON.parse(xhr.responseText);
                }
                catch (e) {
                    onError(e);
                }
                if (res)
                    onSuccess(res);
            }
            else {
                try {
                    res = JSON.parse(xhr.responseText);
                }
                catch (e) {
                    onError(e);
                }
                if (res)
                    onError(res);
            }
        }
    };
    xhr.send(JSON.stringify(body));
}
;
function postRememberancesByIdUpvote(id, onSuccess, onError) {
    var xhr = new XMLHttpRequest();
    xhr.open('POST', '/rememberances/' + encodeURIComponent(id.toString()) + '/upvote', true);
    xhr.setRequestHeader('Accept', 'application/json');
    xhr.onreadystatechange = function () {
        var res = null;
        if (xhr.readyState === 4) {
            if (xhr.status === 204 || xhr.status === 205) {
                onSuccess();
            }
            else if (xhr.status >= 200 && xhr.status < 300) {
                try {
                    res = JSON.parse(xhr.responseText);
                }
                catch (e) {
                    onError(e);
                }
                if (res)
                    onSuccess(res);
            }
            else {
                try {
                    res = JSON.parse(xhr.responseText);
                }
                catch (e) {
                    onError(e);
                }
                if (res)
                    onError(res);
            }
        }
    };
    xhr.send(null);
}
;
function postRememberancesByIdDownvote(id, onSuccess, onError) {
    var xhr = new XMLHttpRequest();
    xhr.open('POST', '/rememberances/' + encodeURIComponent(id.toString()) + '/downvote', true);
    xhr.setRequestHeader('Accept', 'application/json');
    xhr.onreadystatechange = function () {
        var res = null;
        if (xhr.readyState === 4) {
            if (xhr.status === 204 || xhr.status === 205) {
                onSuccess();
            }
            else if (xhr.status >= 200 && xhr.status < 300) {
                try {
                    res = JSON.parse(xhr.responseText);
                }
                catch (e) {
                    onError(e);
                }
                if (res)
                    onSuccess(res);
            }
            else {
                try {
                    res = JSON.parse(xhr.responseText);
                }
                catch (e) {
                    onError(e);
                }
                if (res)
                    onError(res);
            }
        }
    };
    xhr.send(null);
}
;
function listRememberances(limit, offset, container) {
    getRememberances(limit, offset, function (rs) {
        rs.forEach(function (r) {
            var n = createRememberanceElement(r);
            container.appendChild(n);
        });
    }, function (e) {
        throw new Error(e);
    });
}
function listRememberancesOnto(containerId) {
    var container = document.getElementById(containerId);
    var limit = parseInt(container.getAttribute("data-limit"));
    var offset = parseInt(container.getAttribute("data-offset"));
    listRememberances(limit, offset, container);
}
function createRememberanceElement(r) {
    var outside = document.createElement('div');
    outside.id = "out" + r.key.toString();
    outside.className = 'rememberanceHolder';
    var n = document.createElement('div');
    n.id = "r" + r.key.toString();
    n.className = "rememberance";
    var body = document.createElement('div');
    body.textContent = r.val.body;
    body.className = "rText";
    n.appendChild(body);
    var upvote = document.createElement('a');
    upvote.id = "r" + r.key.toString() + "Upvote";
    upvote.className = "upvote";
    upvote.textContent = "+" + r.val.upvotes.toString();
    upvote.title = "Click to Upvote";
    upvote.href = "#";
    upvote.onclick = function (e) {
        postRememberancesByIdUpvote(r.key, function () {
            var upvotes = parseInt(upvote.textContent.slice(1));
            upvote.textContent = "+" + (upvotes + 1).toString();
        }, function (e) { throw Error(e); });
    };
    n.appendChild(upvote);
    var downvote = document.createElement('a');
    downvote.id = "r" + r.key.toString() + "Upvote";
    downvote.className = "downvote";
    downvote.textContent = "-" + r.val.downvotes.toString();
    downvote.title = "Click to Downvote";
    downvote.href = "#";
    downvote.onclick = function (e) {
        postRememberancesByIdDownvote(r.key, function () {
            var downvotes = parseInt(downvote.textContent.slice(1));
            downvote.textContent = "-" + (downvotes + 1).toString();
        }, function (e) { throw Error(e); });
    };
    n.appendChild(downvote);
    outside.append(n);
    return outside;
}
function submitRememberanceForm(textBoxId, containerId) {
    var textBox = document.getElementById(textBoxId);
    var body = textBox.value;
    var container = document.getElementById(containerId);
    postRememberances(body, function (id) {
        alert(id);
        textBox.value = "I remember ";
        var r = {
            key: id,
            val: {
                body: body,
                upvotes: 0,
                downvotes: 0
            }
        };
        var n = createRememberanceElement(r);
        container.prepend(n);
    }, function (e) {
        throw new Error(e);
    });
}
//# sourceMappingURL=main.js.map